src/Components/Web/src/Virtualization/Virtualize.cs

```diff
@@ -0,0 +1,146 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.Rendering;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.Virtualization
+{
+    /// <summary>
+    /// Provides functionality for rendering a virtualized, fixed-size list of items.
+    /// </summary>
+    /// <typeparam name="TItem">The <c>context</c> type for the items being rendered.</typeparam>
+    public sealed class Virtualize<TItem> : VirtualizeBase<TItem>, IAsyncDisposable
+    {
+        private const string JsFunctionsPrefix = "Blazor._internal.Virtualize";
+
+        private DotNetObjectReference<Virtualize<TItem>>? _selfReference;
+
+        private int _itemsBefore;
+
+        private int _itemsVisible;
+
+        private ElementReference _spacerBefore;
+
+        private ElementReference _spacerAfter;
+
+        [Inject]
+        private IJSRuntime JSRuntime { get; set; } = default!;
+
+        /// <inheritdoc />
+        protected override async Task OnAfterRenderAsync(bool firstRender)
+        {
+            if (firstRender)
+            {
+                _selfReference = DotNetObjectReference.Create<Virtualize<TItem>>(this);
+                await JSRuntime.InvokeVoidAsync($"{JsFunctionsPrefix}.init", _selfReference, _spacerBefore, _spacerAfter);
+            }
+        }
+
+        /// <inheritdoc />
+        protected override void BuildRenderTree(RenderTreeBuilder builder)
+        {
+            var itemsAfter = Math.Max(0, ItemSource.Count() - _itemsVisible - _itemsBefore);
+
+            builder.OpenElement(0, "div");
+            builder.AddAttribute(1, "style", GetSpacerStyle(_itemsBefore));
+            builder.AddElementReferenceCapture(2, elementReference => _spacerBefore = elementReference);
+            builder.CloseElement();
+
+            builder.OpenRegion(3);
+
+            if (ItemTemplate != null)
+            {
+                var items = IsVirtualized ? ItemSource.Skip(_itemsBefore).Take(_itemsVisible) : ItemSource;
+
+                foreach (var item in items)
+                {
+                    ItemTemplate(item)(builder);
+                }
+            }
+
+            builder.CloseRegion();
+
+            builder.OpenElement(4, "div");
+            builder.AddAttribute(5, "style", GetSpacerStyle(itemsAfter));
+            builder.AddElementReferenceCapture(6, elementReference => _spacerAfter = elementReference);
+
+            builder.CloseElement();
+
+            if (Footer != null)
+            {
+                builder.OpenRegion(7);
+
+                Footer.Invoke(builder);
+
+                builder.CloseRegion();
+            }
+        }
+
+        /// <summary>
+        /// Called when the top spacer becomes visible.
+        /// This method is intended to be invoked only from JavaScript.
```

Comment:
Since the intent is not really to have this as `public`, a pattern worth considering is declaring a type like:

```
public class VirtualizeJsInteropCallbacks
{
    private readonly Virtualize _owner;

    public VirtualizeJsInteropCallbacks(Virtualize owner)
    {
        _owner = owner;
    }

    [JSInvokable]
    public void OnTopSpacerVisible(...)
    {
        _owner.HandleOnTopSpacerVisible(...);
    }
}
```

... where `HandleOnTopSpacerVisible` is `internal`. That way, developers who get a reference to a `Virtualize<T>` (e.g., using `@ref`) don't have any way of calling the OnXyz methods and can't be confused about thinking they should.